examples/app/flight_delays/main.py

```diff
@@ -0,0 +1,144 @@
+import os
+import datetime
+import numpy as np
+import pandas as pd
+
+from bokeh.models import HBox, VBox
+from bokeh.io import curdoc
+from bokeh.plotting import Figure
+from bokeh.models.widgets import Select, Slider
+from bokeh.models import ColumnDataSource, HoverTool, Paragraph
+from bokeh.palettes import Reds5
+
+import itertools
+
+from bokeh.sampledata.us_states import data as states
+
+def load_data():
+
+    # load states
+    states.pop("HI", None)
+    states.pop("AK", None)
+    state_xs = [states[code]["lons"] for code in states]
+    state_ys = [states[code]["lats"] for code in states]
+    x_chain = list(itertools.chain(*state_xs))
+
+    max_long = max(x_chain)
+    min_long = min(x_chain)
+
+    # load airlines delays
+    airline_delays_df = pd.read_csv(os.path.join(os.path.dirname(__file__), 'data/airline_delay_causes.csv'))
+    airline_delays_df['period'] = airline_delays_df.apply(lambda row: datetime.date(month=row[' month'], year=row['year'], day=1), axis=1)
+    airline_delays_df = airline_delays_df[['period', 'airport', 'arr_flights', 'arr_del15', 'carrier_name']]
+
+    # load airport locations
+    adf = pd.read_csv(os.path.join(os.path.dirname(__file__), 'data/airports.csv'))
+    adf = adf[adf['country'] == 'United States'][['iata/faa', 'latitude', 'longitude', 'name']]
+    adf = adf[(adf.longitude < max_long) & (adf.longitude > min_long)]
+    adf.set_index('iata/faa', inplace=True)
+
+    return state_xs, state_ys, airline_delays_df, adf
+
+def on_year_change(attr, old, new):
+    global year
+    year = new
+    update_map()
+
+def on_month_change(attr, old, new):
+    global month
+    month = new
+    update_map()
+
+def on_carrier_change(attr, old, new):
+    global carrier
+    carrier = new
+    update_map()
+
+def update_map():
+    global p
+    p.title = 'US Airport Delays ({}/{})'.format(month, year)
+
+    date_value = datetime.date(year=year, month=month, day=1)
+    df1 = airline_delays_df[airline_delays_df.period == date_value]
+
+    if carrier != 'All':
+        df1 = df1[df1.carrier_name == carrier]
+
+    df1 = df1.groupby('airport', as_index=False).sum()
+    df1.set_index('airport', inplace=True)
+
+    if df1.empty:
+        data = {}
+        data['x'] = []
+        data['y'] = []
+        source.data = data
+        return
+
+    df1['late_pct'] = df1['arr_del15'] / df1['arr_flights'] * 100
+    df1 = df1[np.isfinite(df1.late_pct)]
+
+    colors = list(reversed(Reds5))
```

Comment:
This work doesn't need to be done every update, suggest making it a module scope variable